File: studio/preview.py
"""Dev preview server manager for studio runs."""

import os
import socket
import subprocess
import time
import urllib.request

from harness import runner

from studio import projects

CREATE_NEW_PROCESS_GROUP = getattr(subprocess, "CREATE_NEW_PROCESS_GROUP", 0)
PREVIEW_TAG_PREFIX = "preview-"
_ACTIVE = {}


def _find_free_port(project):
    lo, hi = project.get("dev_port_range", [4310, 4399])
    for port in range(lo, hi + 1):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            try:
                s.bind(("127.0.0.1", port))
                return port
            except OSError:
                continue
    raise RuntimeError("no free port in dev_port_range")


def _worktree_path(project, run_data):
    cfg = projects.derive_harness_cfg(project, run_data["run_id"])
    wt_root = run_data.get("worktree_root") or cfg["worktree_root"]
    task_id = run_data["task"]["id"]
    return os.path.join(wt_root, task_id)


def _wait_ready(port, timeout_s=60):
    deadline = time.time() + timeout_s
    url = f"http://127.0.0.1:{port}/"
    while time.time() < deadline:
        try:
            with urllib.request.urlopen(url, timeout=2) as resp:
                if resp.status < 600:
                    return True
        except Exception:
            pass
        time.sleep(0.25)
    return False


def start(run_data):
    run_id = run_data["run_id"]
    if run_id in _ACTIVE:
        return _ACTIVE[run_id]
    project = projects.get(run_data["project_id"])
    wt = _worktree_path(project, run_data)
    if not os.path.isdir(wt):
        raise FileNotFoundError(f"worktree not found: {wt}")
    port = _find_free_port(project)
    env = {**os.environ, "PORT": str(port)}
    cmd = project.get("dev_cmd", "npm run dev")
    proc = subprocess.Popen(
        cmd,
        cwd=wt,
        env=env,
        shell=True,
        creationflags=CREATE_NEW_PROCESS_GROUP,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    tag = f"{PREVIEW_TAG_PREFIX}{run_id}"
    runner.ACTIVE.setdefault(tag, set()).add(proc.pid)
    ready = _wait_ready(
        port, timeout_s=float(os.environ.get("STUDIO_PREVIEW_READY_SEC", "60"))
    )
    status = "ready" if ready else "starting"
    info = {"url": f"http://127.0.0.1:{port}/", "port": port, "status": status}
    _ACTIVE[run_id] = {**info, "pid": proc.pid, "tag": tag}
    return info


def get_status(run_id):
    return _ACTIVE.get(run_id, {"url": None, "port": None, "status": "stopped"})


def stop(run_id):
    info = _ACTIVE.pop(run_id, None)
    if info:
        runner.kill_tag(info.get("tag", f"{PREVIEW_TAG_PREFIX}{run_id}"))
    return {"url": None, "port": None, "status": "stopped"}

File: src/server/preview.ts
/** Port of backend/studio/preview.py */
import fs from "node:fs";
import net from "node:net";
import path from "node:path";
import { spawn } from "node:child_process";
import * as legacyProjects from "./legacyProjects";
import { killTag, trackPid } from "./runner";

const PREVIEW_TAG_PREFIX = "preview-";
const ACTIVE = new Map<string, PreviewInfo>();

export type PreviewInfo = {
  url: string;
  port: number;
  status: string;
  pid?: number;
  tag?: string;
};

async function findFreePort(project: legacyProjects.LegacyProject): Promise<number> {
  const range = (project.dev_port_range as number[] | undefined) ?? [4310, 4399];
  const [lo, hi] = range;
  for (let port = lo; port <= hi; port++) {
    const free = await new Promise<boolean>((resolve) => {
      const srv = net.createServer();
      srv.once("error", () => resolve(false));
      srv.once("listening", () => srv.close(() => resolve(true)));
      srv.listen(port, "127.0.0.1");
    });
    if (free) return port;
  }
  throw new Error("no free port in dev_port_range");
}

function worktreePath(project: legacyProjects.LegacyProject, runData: Record<string, unknown>): string {
  const cfg = legacyProjects.deriveHarnessCfg(project, String(runData.run_id));
  const wtRoot = String((runData as { worktree_root?: string }).worktree_root ?? cfg.worktree_root);
  const taskId = String((runData.task as { id: string }).id);
  return path.join(wtRoot, taskId);
}

async function waitReady(port: number, timeoutS = 60): Promise<boolean> {
  const deadline = Date.now() + timeoutS * 1000;
  while (Date.now() < deadline) {
    try {
      const res = await fetch(`http://127.0.0.1:${port}/`, { signal: AbortSignal.timeout(2000) });
      if (res.status < 600) return true;
    } catch {
      /* retry */
    }
    await new Promise((r) => setTimeout(r, 250));
  }
  return false;
}

export async function start(runData: Record<string, unknown>): Promise<PreviewInfo> {
  const runId = String(runData.run_id);
  const existing = ACTIVE.get(runId);
  if (existing) return existing;

  const project = legacyProjects.get(String(runData.project_id));
  if (!project) throw new Error("project not found");
  const wt = worktreePath(project, runData);
  if (!fs.existsSync(wt)) throw new Error(`worktree not found: ${wt}`);

  const port = await findFreePort(project);
  const cmd = String(project.dev_cmd ?? "npm run dev");
  const proc = spawn(cmd, {
    cwd: wt,
    env: { ...process.env, PORT: String(port), CI: "true" },
    shell: true,
    detached: true,
    stdio: "ignore",
    windowsHide: true,
  });
  proc.unref();
  const tag = `${PREVIEW_TAG_PREFIX}${runId}`;
  if (proc.pid) trackPid(tag, proc.pid);

  const readySec = parseFloat(process.env.STUDIO_PREVIEW_READY_SEC ?? "60");
  const ready = await waitReady(port, readySec);
  const info: PreviewInfo = {
    url: `http://127.0.0.1:${port}/`,
    port,
    status: ready ? "ready" : "starting",
    pid: proc.pid,
    tag,
  };
  ACTIVE.set(runId, info);
  return info;
}

export function getStatus(runId: string): PreviewInfo {
  return ACTIVE.get(runId) ?? { url: null as unknown as string, port: null as unknown as number, status: "stopped" };
}

export function stop(runId: string): PreviewInfo {
  const info = ACTIVE.get(runId);
  ACTIVE.delete(runId);
  if (info?.tag) killTag(info.tag);
  return { url: null as unknown as string, port: null as unknown as number, status: "stopped" };
}
